contracts/modules/TransferManager/VolumeRestrictionTransferManager.sol

```diff
@@ -0,0 +1,241 @@
+    /**
+     * @notice This function returns the signature of configure function
+     */
+    function getInitFunction() public pure returns (bytes4) {
+        return bytes4(0);
+    }
+
+    /**
+     * @notice Return the permissions flag that are associated with Percentage transfer Manager
+     */
+    function getPermissions() public view returns(bytes32[]) {
+        bytes32[] memory allPermissions = new bytes32[](1);
+        allPermissions[0] = ADMIN;
+        return allPermissions;
+    }
+
+    // this takes a userAddress as input, and returns a uint that represents the number of tokens allowed to be withdrawn right now
+    function _currentAllowedAmount(address userAddress) internal view returns (uint) {
+        // get lock up array for this user
+        LockUp[] storage userLockUps = lockUps[userAddress];
+        
+        // we will return the total amount allowed for this user right now, across all their lock ups.
+        uint allowedSum = 0;
+
+        // save the total number of granted tokens, ever
+        // so that we can subtract the already withdrawn balance from the amount to be allowed to withdraw
+        uint totalSum = 0;
+        
+        // loop over the user's lock ups
+        for (uint i = 0; i < userLockUps.length; i++) {
+            LockUp storage aLockUp = userLockUps[i];
+
+            // has lockup started yet?
+            if (now < aLockUp.startTime) {
+                // it has not.  don't let them transfer any tokens.
+                continue;
+            }
+
+            // add total amount to totalSum
+            totalSum = totalSum.add(aLockUp.totalAmount);
+
+            // check if lockup has entirely passed
+            if (now >= aLockUp.startTime.add(aLockUp.lockUpPeriodSeconds)) {
+                // lockup has passed, or not started yet.  allow all.
+                allowedSum = allowedSum.add(aLockUp.totalAmount);
+            } else {
+                // lockup is still active. calculate how many to allow to be withdrawn right now
+                
+                // calculate how many periods have elapsed already
+                uint elapsedPeriods = (now.sub(aLockUp.startTime)).div(aLockUp.releaseFrequencySeconds);
+                // calculate the total number of periods, overall
+                uint totalPeriods = aLockUp.lockUpPeriodSeconds.div(aLockUp.releaseFrequencySeconds);
+                // calculate how much should be released per period
+                uint amountPerPeriod = aLockUp.totalAmount.div(totalPeriods);
+                // calculate the number of tokens that should be released,
+                // multiplied by the number of periods that have elapsed already
+                // and add it to the total allowedSum
+                allowedSum = allowedSum.add(amountPerPeriod.mul(elapsedPeriods));
+            }
+        }
+
+        // subtract already withdrawn amount
+        uint currentUserBalance = ISecurityToken(securityToken).balanceOf(userAddress);
```

Comment:
There is a high chance of mistake in calculation. there are many other possibilities where investor/Affiliate balance can change. I strongly suggest adding the mechanism which will keep track of withdrawn balance from the lockups.

Reply:
I understand, but, I'm not sure that the already withdrawn amount should be stored in this transfer module: 

1. verifyTransfer() is public.  it probably shouldn't mutate state internally, right?  
2. there's no way to know inside a transfer module if entire transfer succeeds, right?  another transfer manager might block the transfer.  Or, is there a success callback of some kind?

Maybe there is another place to store this already withdrawn amount?  On the security token itself?

Reply:
The `ManualApprovalTransferManager` does this if you want a concrete example to look at.

Reply:
@adamdossa okay, that definitely resolves my 1st point.  Thanks for explaining!

But I think my second point still stands, right?  What if a token has two transfer managers on it, and this Volume Restriction manager returns Result.VALID and marks a user as having withdrawn tokens.  But the next transfer manager returns Result.INVALID.  Therefore, the transfer never actually happens, but the Volume Restriction manager marked it as happening.  There's now a mismatch between the real world "amount withdrawn" and the amount stored in this Volume Restriction manager.

That's why I asked if there's a success callback of some kind that can tell the Volume Restriction module that the entire transfer succeeded, but I don't see one.  I don't think it would be that hard to add, though.

@satyamakgec or @CPSTL - any thoughts here?

Reply:
If an actual transfer is happening (`_isTransfer == true`) and the transfer fails (i.e. someone returns INVALID) then the whole transaction is reverted, so any state updates you make would be reverted as well.

Reply:
In other words you can assume that if the transfer doesn't actually happen, the whole transfer would be reverted.

Reply:
K, thanks.  Makes sense.  I made those changes.